src/observe.ts

```diff
@@ -0,0 +1,330 @@
+import { getPropertyDescriptor, isIdentical } from './lang';
+import { queueMicroTask } from './queue';
+import Scheduler from './Scheduler';
+
+export class BaseObjectObserver {
+	protected _listener: (events: PropertyEvent[]) => any;
+	protected _propertyStore: {};
+	protected _target: any;
+
+	constructor(kwArgs?: KwArgs) {
+		this._listener = kwArgs.listener;
+		this._propertyStore = {};
+		this._target = kwArgs.target;
+	}
+}
+
+export interface KwArgs {
+	listener: (events: PropertyEvent[]) => any;
+	nextTurn?: boolean;
+	onlyReportObserved?: boolean;
+	target: {};
+}
+
+module ES5 {
+	/**
+	* The scheduler used to mimic native `Object.observe` reporting. Note that all changes to all
+	* observed objects are notified within the same scheduled microtask.
+	*/
+	const scheduler = new Scheduler({ queueFunction: queueMicroTask });
+
+	export class Observer extends BaseObjectObserver implements Observer {
+		/**
+		* Determines whether change notifications should be fired immediately (`false`) or queued with
+		* the scheduler (`true`, thus mimicking native `Object.observe` behavior). Defaults to `true`.
+		*/
+		nextTurn: boolean;
+
+		protected _boundDispatch: () => void;
+		protected _currentlyScheduled: Hash<PropertyEvent>;
+		protected _descriptors: Hash<PropertyDescriptor>;
+
+		/**
+		* Creates a new Observer to watch and notify listeners of changes.
+		*
+		* This should only be used when 1) there is no native `Object.observe` implementation or 2) notifications
+		* should be fired immediately rather than queued.
+		*
+		* @constructor
+		*
+		* @param kwArgs
+		* The `kwArgs` object is expected to contain the target object to observe and the callback
+		* that will be fired when changes occur.
+		*/
+		constructor(kwArgs: KwArgs) {
+			super(kwArgs);
+
+			this.nextTurn = ('nextTurn' in kwArgs) ? kwArgs.nextTurn : true;
+
+			this._descriptors = {};
+			this._boundDispatch = this._dispatch.bind(this);
+		}
+
+		protected _dispatch() {
+			let queue = this._currentlyScheduled;
+			let events: PropertyEvent[] = Object.keys(queue).map(function (property: string): PropertyEvent {
+				return queue[property];
+			});
+
+			this._currentlyScheduled = null;
+			this._listener(events);
+		}
+
+		protected _restore(property: string): void {
+			let target = this._target;
+			let store = this._propertyStore;
+
+			Object.defineProperty(target, property, (this._descriptors[property] || {
+				configurable: true,
+				enumerable: true,
+				value: target[property],
+				writable: true
+			}));
+
+			target[property] = (<any> store)[property];
+		}
+
+		protected _schedule(property: string): void {
+			let event: PropertyEvent = {
+				target: this._target,
+				name: property
+			};
+
+			if (this.nextTurn) {
+				if (!this._currentlyScheduled) {
+					this._currentlyScheduled = {};
+					scheduler.schedule(this._boundDispatch);
+				}
+
+				this._currentlyScheduled[property] = event;
+			}
+			else {
+				this._listener([ event ]);
```

Comment:
Can we add a test to cover this scenario?

The test should create an observer with nextTick = false; observe a property and be notified of a change synchronously before the test finishes.